src/jupyterhub_config/custom_spawner.py

```diff
@@ -211,3 +198,25 @@ def _configure_notebook_dir(self, username: str, user_dir: Path):
         else:
             self.log.info(f'Non-admin user detected: {username}. Setting up user-specific workspace.')
             self.notebook_dir = str(user_dir)
+
+    def _ensure_workspace_permission(self, user_dir: Path, username: str):
+        """
+        Ensure the user's workspace has the correct permissions.
+        """
+        try:
+            user_info = pwd.getpwnam(username)
+        except KeyError:
+            raise ValueError(f'System user {username} does not exist')
+        gid = user_info.pw_gid
+        group_name = grp.getgrgid(gid).gr_name
+
+        self.log.info(f'Configuring workspace permissions for {username}')
+        # Change the directory's ownership to the user
+        subprocess.run(['sudo', 'chown', '-R', f'{username}:{group_name}', user_dir], check=True)
+
+        self.log.info(f'Add spark_user to the group of {group_name}')
+        subprocess.run(['sudo', 'usermod', '-aG', group_name, 'spark_user'], check=True)
+
+        # TODO: Set directory permissions to 750 or 700 or switch to use docker spawner
+        # Set directory permissions to 777: Owner (rwx), Group (rwx), Others (rwx)
+        subprocess.run(['sudo', 'chmod', '-R', '777', user_dir], check=True)
```

Comment:
After exploring DockerSpawner, I'm confident that switching to it will resolve the issue, as each user's directory is mounted individually per container.